Examples/Scripts/IcaFixProcessingBatch.sh

```diff
@@ -157,53 +157,68 @@ main() {
 	# establish queue for job submission
 	QUEUE="-q hcp_priority.q"
 	if [ "${RunLocal}" == "TRUE" ]; then
-		queuing_command=""
+		queuing_command=()
 	else
-		queuing_command="${FSLDIR}/bin/fsl_sub ${QUEUE}"
+		queuing_command=("${FSLDIR}/bin/fsl_sub" "${QUEUE}")
 	fi
 
-	DIR=$(pwd)
-	
 	for Subject in ${Subjlist}; do
 		echo ${Subject}
 
 		ResultsFolder="${StudyFolder}/${Subject}/MNINonLinear/Results"
-		cd ${ResultsFolder}
 		
-		if [ -z ${ConcatName} ]; then
+		if [ -z ${ConcatNames} ]; then
 			# single-run FIX
 			FixScript=${HCPPIPEDIR}/ICAFIX/hcp_fix
 			
-			for fMRIName in ${fMRINames}; do
+			fMRINamesFlat=$(echo ${fMRINames} | sed 's/@/ /g')
+			
+			for fMRIName in ${fMRINamesFlat}; do
 				echo "  ${fMRIName}"
 
-				InputFile="${fMRIName}/${fMRIName}"
+				InputFile="${ResultsFolder}/${fMRIName}/${fMRIName}"
 
-				cmd="${queuing_command} ${FixScript} ${InputFile} ${bandpass} ${domot} ${TrainingData} ${FixThreshold}"
-				echo "About to run: ${cmd}"
-				${cmd}
+				cmd=("${queuing_command[@]}" "${FixScript}" "${InputFile}" ${bandpass} ${domot} "${TrainingData}" ${FixThreshold})
+				echo "About to run: ${cmd[*]}"
+				"${cmd[@]}"
 			done
 
 		else
-			# multi-run FIX
-			FixScript=${HCPPIPEDIR}/ICAFIX/hcp_fix_multi_run
-			ConcatNameFile="${ConcatName}/${ConcatName}"
-
-			InputFile=""			
-			for fMRIName in ${fMRINames}; do
-				InputFile+="${fMRIName}/${fMRIName}@"
-			done
+        	#need arrays to sanity check number of concat groups
+        	IFS=' ' read -a concatarray <<< "${ConcatNames}"
+        	IFS=@ read -a fmriarray <<< "${fMRINames}"
+        	
+        	if ((${#concatarray[@]} != ${#fmriarray[@]})); then
+        	    echo "ERROR: number of names in ConcatNames does not match number of fMRINames groups"
+        	    exit 1
+        	fi
+
+		    for ((i = 0; i < ${#concatarray[@]}; ++i))
+		    do
+		        ConcatName="${ResultsFolder}/${concatarray[$i]}/${concatarray[$i]}.nii.gz"
+		        fMRINamesGroup="${fmriarray[$i]}"
+			    # multi-run FIX
+			    FixScript=${HCPPIPEDIR}/ICAFIX/hcp_fix_multi_run
+			    ConcatFileName="${ConcatName}/${ConcatName}"
+
+			    InputFile=""
+			    for fMRIName in ${fMRINamesGroup}; do
+			        if [[ "$InputFile" == "" ]]; then
+			            InputFile="${ResultsFolder}/${fMRIName}/${fMRIName}"
+		            else
+    				    InputFile+="@${ResultsFolder}/${fMRIName}/${fMRIName}"
+				    fi
+			    done
 			
-			echo "  InputFile: ${InputFile}"
+			    echo "  InputFile: ${InputFile}"
 
-			cmd="${queuing_command} ${FixScript} ${InputFile} ${bandpass} ${ConcatFileName} ${domot} ${TrainingData} ${FixThreshold}"
-			echo "About to run: ${cmd}"
-			${cmd}
+			    cmd=("${queuing_command[@]}" "${FixScript}" "${InputFile}" ${bandpass} ${ConcatFileName} ${domot} "${TrainingData}" ${FixThreshold})
+			    echo "About to run: ${cmd[*]}"
```

Comment:
This is for my curiosity -- why do you use `${cmd[*]}` in the echo statement, but `${cmd[@]}` to actually invoke the command -- what's the difference?

Reply:
Partly habit - the logging functions at least used to only use $1, so if you passed them multiple arguments only the first one would be used - using @ within "" is special, it expands to multiple arguments, while * expands to a single argument with spaces between elements (if the default IFS is in effect):

```
tim@timsdev:~$ test=(a b c)
tim@timsdev:~$ echo "testing ${test[@]}"
testing a b c
tim@timsdev:~$ function countargs() { echo $#; }
tim@timsdev:~$ countargs "testing ${test[@]}"
3
tim@timsdev:~$ countargs "testing ${test[*]}"
1
```

Reply:
That's a helpful example.  But just to make sure I understand, in this particular case, since `"${cmd[@]}"` includes both the command itself, and its arguments (i.e., we aren't actually passing that into a function/script), you would get the same result if you used `"${cmd[*]}"` to invoke the command, right?  Conversely, for the echo statement, `echo "About to run: ${cmd[@]}"` would work, right?

Reply:
For specifically echo, yes, as shown in the example, it does work to use @.  The result of echo is just a string (like any other command's output), and echo uses every argument you give it.  However, `"${cmd[*]}" would not work, as it would try to run a massively long executable name that contains spaces in it, as it would put all the arguments into a single lump with the actual intended executable name, consider:

```
tim@timsdev:~$ echo hi
hi
tim@timsdev:~$ "echo hi"
echo hi: command not found
```

Executing commands in bash is NOT just based on a simple string (side note, windows command line actually does just pass a single string, and each executable has to parse that string in order to pretend that it was being passed multiple arguments - posix exec functions use an array of strings to start an executable, so its arguments truly are separate at the lowest level, and that separation is under the control of the shell itself).  Bash needs to know that each element of the array should be used separately (first element is the executable, second is the first argument, etc).

Using @ or * as an index inside double quotes is the same as the difference between `"$@"` and `"$*"`: `"$@"` expands to `"$1" "$2" "$3"`, etc, each as separate parts of the command, while `"$*"` expands to `"$1 $2 $3"`, all passed as a single argument containing spaces.

Using a bare `${cmd[*]}` without quotes would work to run the command as long as there aren't any spaces in any arguments, since without quotes, bash splits on whitespace after variable expansion.  `"${cmd[@]}"` is safe even when the arguments contain internal whitespace, because expansion inside quotes occurs after whitespace splitting.

Reply:
Thanks. Very helpful.  So, all the instances in the scripts in which we use `main $@` should be `main "$@"` instead, to be "whitespace safe"?

Reply:
Yes, that would be better.

Reply:
Note, though, that if quotes aren't used around variable names containing files inside main itself when executing commands, spaces inside arguments will still break scripts, it will just cause commands to fail, rather than the arguments to main being split incorrectly.